tests/trainer/test_sharded_checkpoint.py

```diff
@@ -0,0 +1,183 @@
+    state_dict2_optim_params = state_dict2['optimizers']['Adam']['state']
+    state_dict1_keys = set(state_dict1_optim_params.keys())
+    state_dict2_keys = set(state_dict2_optim_params.keys())
+    assert len(state_dict1_keys.symmetric_difference(state_dict2_keys)) == 0, textwrap.dedent(
+        f"""The two state dicts being compared must have the exact same set of keys,
+        but instead these keys that belong to one, but not the other:
+        {state_dict1_keys.symmetric_difference(state_dict2_keys)}""")
+
+    for param_name in state_dict2_optim_params.keys():
+        state_dict1_param_moment_dict = state_dict1_optim_params[param_name]
+        state_dict2_param_moment_dict = state_dict2_optim_params[param_name]
+        for moment_name in state_dict2_param_moment_dict.keys():
+            state_dict1_moment = state_dict1_param_moment_dict[moment_name]
+            state_dict2_moment = state_dict2_param_moment_dict[moment_name]
+            assert torch.equal(
+                state_dict1_moment,
+                state_dict2_moment), f'Moment {moment_name} for parameter {param_name} not the same between state dicts'
+
+
+def _compare_model_params_between_state_dicts(state_dict1, state_dict2):
+    # Check that model params are equal between in memory mode and checkpoint
+    state_dict1_model_params = state_dict1['model']
+    state_dict2_model_params = state_dict2['model']
+
+    state_dict1_keys = set(state_dict1_model_params.keys())
+    state_dict2_keys = set(state_dict2_model_params.keys())
+    assert len(state_dict1_keys.symmetric_difference(state_dict2_keys)) == 0, textwrap.dedent(
+        f"""The two state dicts being compared must have the exact same set of keys,
+        but instead these keys that belong to one, but not the other:
+        {state_dict1_keys.symmetric_difference(state_dict2_keys)}""")
+
+    for param_name in state_dict2_model_params.keys():
+        state_dict1_model_tensor = state_dict1_model_params[param_name]
+        state_dict2_model_tensor = state_dict2_model_params[param_name]
+        assert torch.equal(state_dict1_model_tensor,
+                           state_dict2_model_tensor), f'Weight named {param_name} not the same between state_dicts'
+
+
+@pytest.mark.gpu
+@world_size(2)
+def test_fsdp_full_state_dict_save(world_size, tmp_path: pathlib.Path):
+
+    save_folder = tmp_path
+    save_filename = 'rank{rank}.pt'
+    trainer = get_trainer(save_folder=str(save_folder), save_filename=save_filename, fsdp_state_dict_type='full')
+
+    trainer.fit()
+    rankn_checkpoint = save_folder / pathlib.Path(f'rank{dist.get_global_rank()}.pt')
+
+    # Check that rank 0 saves a checkpoint to disk, but rank 1 does not.
+    if dist.get_global_rank() == 0:
+        assert os.path.exists(rankn_checkpoint)
+    elif dist.get_global_rank() == 1:
+        assert not os.path.exists(rankn_checkpoint)
+    state_dict_in_memory = trainer.state.state_dict()
```

Comment:
no I think that will just return rank 1's shard. This is because we set `rank0_only=True`. See this code I added to state.py: 
```python
if state_dict_type == 'full':
        state_dict_config = FullStateDictConfig(offload_to_cpu=True, rank0_only=True)
        fsdp_state_dict_type = StateDictType.FULL_STATE_DICT
```

Reply:
ah ok, in that case, maybe the test could confirm that?

Reply:
ok added a test to check the rank 1 state dict